internal/v4/archive.go

```diff
@@ -342,24 +367,100 @@ func checkCharmIsValid(ch charm.Charm) error {
 	m := ch.Meta()
 	for _, rels := range []map[string]charm.Relation{m.Provides, m.Requires, m.Peers} {
 		if err := checkRelationsAreValid(rels); err != nil {
-			return errgo.Mask(err)
+			return errgo.Mask(err, errgo.Is(params.ErrInvalidEntity))
 		}
 	}
+	if err := checkConsistentSeries(m.Series); err != nil {
+		return errgo.Mask(err, errgo.Is(params.ErrInvalidEntity))
+	}
 	return nil
 }
 
 func checkRelationsAreValid(rels map[string]charm.Relation) error {
 	for _, rel := range rels {
 		if rel.Name == "relation-name" {
-			return errgo.Newf("relation %s has almost certainly not been changed from the template", rel.Name)
+			return errgo.WithCausef(nil, params.ErrInvalidEntity, "relation %s has almost certainly not been changed from the template", rel.Name)
 		}
 		if rel.Interface == "interface-name" {
-			return errgo.Newf("interface %s in relation %s has almost certainly not been changed from the template", rel.Interface, rel.Name)
+			return errgo.WithCausef(nil, params.ErrInvalidEntity, "interface %s in relation %s has almost certainly not been changed from the template", rel.Interface, rel.Name)
 		}
 	}
 	return nil
 }
 
+type distribution string
```

Comment:
the next PR will formalize this a bit and add an appropriate docstring, so I'll leave this for now.